contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,857 @@
+}
+```
+
+## Types and Interfaces
+
+Use **type** by default.
+
+```ts
+// Bad
+interface Report {
+    // properties
+}
+
+// Good
+type Report = {
+    // properties
+}
+```
+
+TypeScript supports both type and interface expressions. Types can be used more broadly: to name primitives, unions, objects and any other type. As these two forms are practically equivalent, it's best to select one over the other to maintain consistency and reduce variation.
+
+One drawback of using interfaces is that they can be declared multiple times and are merged. This can lead to unintended behavior.
+
+```ts
+// Merging interfaces
+interface Foo {
+  a: number;
+}
+interface Foo {
+  b: number;
+}
+
+// Foo is now { a: number; b: number; }
+```
+
+Use interfaces only when you encounter a situation where `type` does not fulfill your requirements. Example scenario is when you need to modify or extend an interface from a third-party library that lacks typings or has incorrect typings. 
+
+When extending or correcting interfaces from third-party libraries, always use [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) to maintain code consistency and avoid potential issues.
+
+```ts
+// Third-party library with missing or incorrect typings
+interface LibraryComponentProps {
+  // Existing incorrect or incomplete typings
+}
+
+// Your code
+declare module 'library/path' {
+  interface LibraryComponentProps {
+    // Add or modify the typings
+    additionalProp: string;
+  }
+}
+```
+
+## Enums and Union Types
+
+Use **union types** instead of enums. Opt for union types whenever possible. Union types provide flexibility and can handle varying data types and dynamic value sets.
+
+```ts
+// Bad
+enum Colors {
+  Red,
+  Green,
+  Blue
+}
+
+// Good
+type Colors = 'red' | 'green' | 'blue';
+```
+You can also define a union type by deriving the type from an array. It's a great choice when seeking an easily iterable and directly usable array in code while keeping one source of truth.
+
+```ts
+// Bad
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { TupleToUnion } from 'type-fest';
+
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = TupleToUnion<typeof COLORS>;
+```
+
+Defining union types using an object offers several benefits. This approach it's a great choice when **enum-like identifiers are needed**, while maintaining one source of truth and improving readability.
+
+```ts
+// Bad
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { ValueOf } from 'type-fest';
+
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = ValueOf<typeof COLORS]>
+```
+
+If using enum is necessary, utilize string enum. Do not use numeric and heterogenous enums. This can help prevent potential bugs at runtime.
```

Comment:
> This can help prevent potential bugs at runtime.

How, why? I'm curious.

Reply:
Some examples:

```ts
enum NumberEnum {
  A, // when not assigning anything, the value will be 0.
  B, // and will increment, 1.
  C, // 2...
}

enum NumberEnum {
  A = 10, // 10
  B, // 11
  C, // 12
  D = 200, // 200
  E, // now increments to 201, 202, 203...
}

enum HeterogenousEnum {
    VALUE_1 = 'string value', // "string value"
    VALUE_2 = 0, // 0
    VALUE_3, // 1
    VALUE_4 = 200, // 200
    VALUE_5, // 201
    VALUE_6 = 'string value',
    VALUE_7 = 0, // 0
    VALUE_8 // 1
}
```

Also, when using numeric enums we would have reverse mapping, for example:

```ts
enum NumberEnum {
  A = 10, // 10.
  B, // 11.
  C, // 12...
}

console.log(NumberEnum.A); // outputs 10.
console.log(NumberEnum[10]); // outputs "A".
```

There are just some examples of how dangerous using numeric/heterogenous enum can be.

Reply:
Awesome thanks. Could you please link to some examples like these or add a brief description for why numeric and heterogenous enums are dangerous? Something like "numeric enums are dangerous because they might not increment as expected, and heterogenous enums are dangerous because the type of a given value is unclear."?